hpke-double-ext.md

```diff
@@ -0,0 +1,210 @@
+an asymmetric key or an asymmetric key and PSK are left unspecified.
+
+
+## Three Party Encryption Context
+
+The sender can create its encryption context in the same way as in 
+standard HPKE:
+
+    def SetupDoubleS(pkR, info):
+        shared_secret, enc = Encap(pkR)
+        return enc, KeySchedule(mode_double, shared_secret, info,
+                                default_psk, default_psk_id)
+
+For the relay, context generation is the same as for the sender
+
+    def SetupDoubleX(pkR, info):
+        shared_secret, enc = Encap(pkR)
+        return enc, KeySchedule(mode_double, shared_secret, info,
+                                default_psk, default_psk_id)
+
+The receiver has to keep track of both key schedules now:
+
+    def SetupDoubleR(encS, encX, skR, info):
+        shared_secret_S = Decap(encS, skR)
+        shared_secret_X = Decap(encX, skR)
+    
+        key_schedule_S = KeySchedule(mode_double, shared_secret_S, info,
+                                default_psk, default_psk_id)
+    
+        key_schedule_R = KeySchedule(mode_double, shared_secret_R, info,
+                                default_psk, default_psk_id)
+    
+        return (key_schedule_S, key_schedule_R)
+
+## Encryption, Double Encryption and Decryption
+
+Creation of level-1 ciphertexts, i.e. regular encryption, from sender to receiver is exactly the same as in standard HPKE:
+
+    def ContextS.Seal(aad, pt):
+        ct = Seal(self.key, self.ComputeNonce(self.seq), aad, pt)
+        self.IncrementSeq()
+        return ct
+
+Level-2 ciphertexts are created by encrypting a level-1 ciphertext one more time using the relays encryption context:
+
+    def ContextX.ReSeal(aad, ct):
+        if ct_level(ct) != 1:
+            raise InvalidParameters
+        cct = Seal(self.key, self.ComputeNonce(self.seq), aad, ct)
+        self.IncrementSeq()
+        return cct
+
+Decryption is only defined on level-2 ciphertexts:
+
+    def ContextR.Open(aad, cct):
+        if ct_level(cct) != 2:
+            raise InvalidParameters
+        ct = Open(self.key, self.ComputeNonce(self.seq), aad, cct)
+        pt = Open(self.key, self.ComputeNonce(self.seq), aad, ct)
+        if pt == OpenError:
+          raise OpenError
+        self.IncrementSeq()
+        return pt
+        
+## Single-Shot Double HPKE
+HPKE Double encryption can already be implemented using the single-shot
+basic mode API provided by standard HPKE. We specify the following
+serialization/deserialization scheme for HPKE ciphertexts:
+
+```text
+    def SerializeHPKECt(enc, ct):
+        return I2OSP(len(enc), 4) || I2OSP(len(ct), 4) || enc || ct
+        
+    def DeserializeHPKECt(bytes):
+        len_enc = OS2IP(bytes[0..4])
+        len_ct = OS2IP(bytes[4..8])
+        return (enc = bytes[8..8 + len_enc], ct = bytes [8 + len_enc ... 8 + len_enc + len_ct])
+```
+
+### Level-1 Encryption
+```text
+    def SealDouble(pkR, ptxt):
+        enc, ct = SealBase(pkR, "Level-1", "", ptxt, None, None, None)
```

Comment:
You use the RFC style notation before and start switching to the hacspec functions here. Let's stick with one.
I suggest keep using the RFC style pseudocode here as well and mention the function names in the comment on the code.

Reply:
That makes sense! I have brought it closer to the RFC style and will address it in the code together with the rest of the ScrambleDB documentation.